shipyard_blueprints/tableau/shipyard_tableau/tableau.py

```diff
@@ -1,78 +1,481 @@
 import tableauserverclient as tsc
-from shipyard_templates import DataVisualization, ShipyardLogger
+import requests
+import jwt
+import uuid
+import datetime
+from shipyard_templates import (
+    DataVisualization,
+    ShipyardLogger,
+    ExitCodeException,
+    datavisualization,
+)
+from typing import Optional, Dict, Any
+from .errors import (
+    EXIT_CODE_JOB_STATUS_ERROR,
+    EXIT_CODE_TABLEAU_WORKBOOK_REFRESH_ERROR,
+    EXIT_CODE_TABLEAU_DATASOURCE_REFRESH_ERROR,
+    EXIT_CODE_STATUS_INCOMPLETE,
+    EXIT_CODE_FINAL_STATUS_ERRORED,
+    EXIT_CODE_FINAL_STATUS_CANCELLED,
+    InvalidWorkbookRequest,
+    TableauJWTAuthError,
+    TableauPATAuthError,
+    TableauUNPWDAuthError,
+    InvalidDatasourceRequest,
+    InvalidViewRequest,
+)
 
 logger = ShipyardLogger.get_logger()
 
 
 class TableauClient(DataVisualization):
```

Comment:
I strongly recommend using a class factory here to handle the auth types. This will clean up all of the optional fields and logic flows for the different credential types.

according to our friend Chat GPT the it would look something like

```
import datetime
import jwt
import requests
import uuid
from typing import Optional

class TableauAuth:
    def authenticate(self):
        raise NotImplementedError("Authenticate method not implemented.")

class UsernamePasswordAuth(TableauAuth):
    def __init__(self, server_url: str, username: str, password: str, site: str):
        self.server_url = server_url
        self.username = username
        self.password = password
        self.site = "" if str(site).lower() == "default" else site

    def authenticate(self):
        payload = f"""
            <tsRequest>
              <credentials name="{self.username}" password="{self.password}" >
                <site contentUrl="{self.site}" />
              </credentials>
            </tsRequest>
        """
        url = f"{self.server_url}/api/3.22/auth/signin"
        headers = {"Content-Type": "application/xml", "Accept": "application/json"}
        response = requests.post(url, headers=headers, data=payload)
        if response.ok:
            auth_token = response.json()["credentials"]["token"]
            site_id = response.json()["credentials"]["site"]["id"]
            user_id = response.json()["credentials"]["user"]["id"]
            return auth_token, site_id, user_id
        else:
            raise ValueError(f"Error logging into Tableau with username and password: {response.text}")

class AccessTokenAuth(TableauAuth):
    def __init__(self, server_url: str, token_name: str, token_secret: str, site: str):
        self.server_url = server_url
        self.token_name = token_name
        self.token_secret = token_secret
        self.site = "" if str(site).lower() == "default" else site

    def authenticate(self):
        payload = f"""
            <tsRequest>
              <credentials personalAccessTokenName="{self.token_name}"
                personalAccessTokenSecret="{self.token_secret}" >
                <site contentUrl="{self.site}" />
              </credentials>
            </tsRequest>
        """
        url = f"{self.server_url}/api/3.22/auth/signin"
        headers = {"Content-Type": "application/xml", "Accept": "application/json"}
        response = requests.post(url, headers=headers, data=payload)
        if response.ok:
            auth_token = response.json()["credentials"]["token"]
            site_id = response.json()["credentials"]["site"]["id"]
            user_id = response.json()["credentials"]["user"]["id"]
            return auth_token, site_id, user_id
        else:
            raise ValueError(f"Error logging into Tableau with access token: {response.text}")

class JwtAuth(TableauAuth):
    def __init__(self, server_url: str, client_id: str, secret_id: str, secret_value: str, username: str, site: str):
        self.server_url = server_url
        self.client_id = client_id
        self.secret_id = secret_id
        self.secret_value = secret_value
        self.username = username
        self.site = "" if str(site).lower() == "default" else site

    def authenticate(self):
        token = jwt.encode(
            {
                "iss": self.client_id,
                "exp": datetime.datetime.utcnow() + datetime.timedelta(minutes=5),
                "jti": str(uuid.uuid4()),
                "aud": "tableau",
                "sub": self.username,
                "scp": [
                    "tableau:views:*",
                    "tableau:workbooks:*",
                    "tableau:datasources:*",
                    "tableau:tasks:*",
                    "tableau:users:*",
                    "tableau:projects*",
                    "tableau:metrics:*",
                    "tableau:groups:*",
                    "tableau:sites:*",
                    "tableau:permissions:*",
                    "tableau:jobs:*",
                    "tableau:job:*",
                ],
            },
            self.secret_value,
            algorithm="HS256",
            headers={"kid": self.secret_id, "iss": self.client_id},
        )
        payload = f"""
            <tsRequest>
              <credentials jwt="{token}">
                <site contentUrl="{self.site}" />
              </credentials>
            </tsRequest>
        """
        url = f"{self.server_url}/api/3.22/auth/signin"
        headers = {"Content-Type": "application/xml", "Accept": "application/json"}
        response = requests.post(url, headers=headers, data=payload)
        if response.ok:
            auth_token = response.json()["credentials"]["token"]
            site_id = response.json()["credentials"]["site"]["id"]
            user_id = response.json()["credentials"]["user"]["id"]
            return auth_token, site_id, user_id
        else:
            raise ValueError(f"Error logging into Tableau with JWT: {response.text}")
```

Reply:
Going to push back on this one as well. I agree that the optional fields should be eliminated, but I do not see the benefit in adding 3 separate classes for the different auth types when 3 separate functions would suffice. To me, three distinct classes is overkill and overcomplicating the situation, when we could have separate functions to instantiate based off of the auth type provided